### PR TITLE
Add BIP324: v2 P2P Transport Protocol

### PR DESCRIPTION
**Pull request by dhruv**
Supersedes #1024

Link to rendered version for your convenience: https://github.com/dhruv/bips/blob/bip324/bip-0324.mediawiki

Open questions:
- The BIP text currently includes a [table for 1-byte message type IDs](https://github.com/dhruv/bips/blob/bip324/bip-0324.mediawiki#v2_Bitcoin_P2P_message_structure). The process for updating this table atomically to assign new IDs when new message types are being proposed is likely going to be an annoyance. What are better ways to handle ongoing allocation?

### CODE REVIEW

File: README.mediawiki
Comment (ajtowns) [inline README.mediawiki:990]: The `style="background-color: #ffffcf"` line needs to be after bip-324, not before it.

Comment (dhruv) [reply README.mediawiki]: That fixed it. Thanks, @ajtowns!

File: bip-0324.mediawiki
Comment (ajtowns) [inline bip-0324.mediawiki:558]: `VERSION`, `VERACK`, `SENDHEADERS`, `WTXIDRELAY`, and `SENDADDRV2` are only sent at most once per connection; wouldn't it make sense to reserve single byte message types for things that will give a meaningful benefit from compression?

File: bip-0324.mediawiki
Comment (ajtowns) [inline bip-0324.mediawiki:546]: The [BIP-61](https://github.com/bitcoin/bips/blob/master/bip-0061.mediawiki) `REJECT` message isn't supported by bitcoin core since 0.19.0, but does seem to still be supported by [btcd](https://github.com/btcsuite/btcd/blob/38ee9a41c8f8aa24a079a28f5e8a86faecffdfe1/peer/peer.go#L997) eg, so maybe would be worth including? On the other hand, maybe there's no point making messages that are only useful for debugging more efficient by a few bytes?

File: bip-0324.mediawiki
Comment (ajtowns) [inline bip-0324.mediawiki:534]: Is there any reason to expect clients supporting this BIP to use `ADDR` rather than `ADDRV2`?

File: bip-0324.mediawiki
Comment (ajtowns) [inline bip-0324.mediawiki:512]: Updating this table seems like it might be annoying (both if proposals made at the same time both choose the same byte, and if a proposal using just an ascii string wants to add support for a one byte id, while remaining compatible with nodes that aren't aware of the new id) -- it seems like it would serialize adding features in a similar way to having to bump the protocol version number to indicate support for a feature.

Did you consider a more generic approach, like just running the protocol through a compression stage prior to encryption, and letting that automatically reduce commonly used message types into a shorter byte stream?

File: bip-0324.mediawiki
Comment (dhruv) [reply bip-0324.mediawiki:558]: There are 210 additional message type IDs still available to allocate. In the future, if we did come close to using them up, a future upgrade could interpret 0xFF as "the next byte is also a part of the message type ID". Given that, it made sense to allocate an ID to all message types still in use.

File: bip-0324.mediawiki
Comment (dhruv) [reply bip-0324.mediawiki:512]: In the time that I worked on it, we did not consider such a generic approach. However, I imagine such an approach might come with new dependencies and an increased surface area?

Comment (ajtowns) [reply bip-0324.mediawiki:512]: I agree a custom dictionary doesn't seem workable; I was thinking of having a compression stream over the entire connection lifetime, and invoking `flush` after each message, or perhaps when the uncompressed send buffer is empty.

I guess updating the mapping as a whole (this bip specifies table version 0, then later define table version 1 mapping 13..255 to some particular set of messages, then version 2 later provides some other mapping), and listing out the table versions you both support either using the VERSION message, or via some new message prior to VERACK, and then using the highest version that you both support once VERACK is completed would work. Perhaps message type 0 could be used for VERSION, since there's presumably no way it can be changed while retaining compatibility?

Comment (ajtowns) [reply bip-0324.mediawiki:512]: If you're mapping `44=WTXIDRELAY` then the receiver presumably either knows how to interpret `WTXIDRELAY`, and can map it to a one or two byte enum `MSG_WTXIDRELAY`, or doesn't know how to interpret `WTXIDRELAY` and can map it to the same one/two byte enum as `MSG_INVALID` or `MSG_IGNORE`. So really this would be an up to 3kB message to announce your mappings but only 250 to 500 bytes per peer to actually track the mapping.

**Comment by ProofOfKeags** [comment]
Since TLS isn't being used is there any way to recover "virtual host" functionality? This is necessary for addressing multiple nodes behind the same NAT which is increasingly common these days.

**Comment by yojoe** [comment]
The [Transport layer specification](https://github.com/dhruv/bips/blob/bip324/bip-0324.mediawiki#Transport_layer_specification) says:

> Given a newly-established connection (typically TCP/IP) between two v2 P2P nodes, ...

What does "typically TCP/IP" mean? Will UDP based connections be supported by BIP-324 as well?

I'm asking, because I stumbled over this `obfs4`-like protocol today which supports both, TCP and UDP:
https://lib.rs/crates/sosistab
`sosistab` seems to have similar goals as BIP-324 and seems to be in active development for over a year now. Maybe there are some learnings/takeaways for BIP-324 here? Especially the following aspect of `sosistap` might of interest for Bitcoin:
 > Avoids last-mile congestive collapse but works around lossy links. Shamelessly unfair in permanently congested WANs --- but that's really their problem, not yours.

Any thoughts?

**Comment by dhruv** [comment]
@yojoe The encrypted protocol we are proposing requires state and the order of packets is important. So we require a stream interface and can't use UDP.

**Comment by yojoe** [comment]
> BIP324 assumes a reliable stream interface to run on top of

So BIP324 "could" run on [QUIC](https://en.wikipedia.org/wiki/QUIC) and similar protocols, which use UDP instead of TCP to provide reliable streams over multiplexed connections?

File: bip-0324.mediawiki
Comment (dergoegge) [inline bip-0324.mediawiki:358]: Am I correct in assuming that this inbound downgrade logic only exists because service flags are additive and v1 clients will make connection to addresses that have the v2 service bit?

Could it make more sense to introduce a new BIP155 network type (e.g. `NET_V2`) instead of using the service flags? All the current network types essentially define a separate transport protocol (IPv{4,6}, Tor, I2P, CJDNS) while the current service flags are more about what type of application data a node serves (blocks, witnesses, compact filters, bloom filters).

Comment (mas722) [reply bip-0324.mediawiki:358]: ****

**Comment by dr-orlovsky** [comment]
Bitcoin P2P connections are still asymmetric: one peer connects to the other peer, and that other peer must be known by some id. For now, the connection happens by the network address (which serves that id). The proposed wire protocol doesn't provide any means of authenticating other peer, which opens possibilities for MiTs, and argues that this increases privacy. I see two contradictions:
- the connecting peer knows the peer it connects to; no anonymity here is gained;
- not authenticating the remote peer keeps the possibility for MiT attacks from the current wire protocol.

I think that "exchange-known" key scheme (like in Noise_XK) is more adequate for bitcoin P2P than "exchange-exchange", as in the current proposal (and like in Noise_XX).

**Review by dr-orlovsky — COMMENTED** [review]
Bitcoin P2P connections are asymmetric: one peer connects to the other peer, and that other peer must be known by some id. For now, the connection happens by the network address (which serves that id). The proposed wire protocol doesn't provide any means of authenticating other peer, which opens possibilities for MiTs, and argues that this increases privacy. I see two contradictions:
- the connecting peer knows the peer it connects to; no anonymity here is gained;
- not authenticating the remote peer keeps the possibility for MiT attacks from the current wire protocol.

I think that "exchange-known" key scheme (like in Noise_XK) is more adequate for bitcoin P2P than "exchange-exchange", as in the current proposal (and like in Noise_XX).

**Comment by jackromo888** [comment]
Thanks

**Comment by dr-orlovsky** [comment]
> authentication is only relevant in cases where the initiator makes a deliberate connection to a peer with whom they have some out-of-band trust relation, and doesn't matter for automatic connections

That is exactly the point I am trying to make: I think that even for automatic connections it might be desirable to know the identity of the peer. The requirement to always have a remote pubkey is not that hard to meet: the remote peers are listed either by the config or via P2P messages, and both of them can require to provide the remote identities alongside network address, as it is done in LN.

At the same time I agree that the abstraction of authentication from encryption is a good thing to have. What is not good is to build a new P2P stack for bitcoin without paying attention to authentication.

**Comment by dr-orlovsky** [comment]
Isn't IP address an identity? Or Onion address? Or any other networking address? Having public key next to it is just a second factor: it doesn't add to the leak of a privacy or doesn't add to be "more identifiable" at all; it is just a second factor ensuring that the _existing identity_ is not stolen by a MitM

**Comment by dr-orlovsky** [comment]
I think there is terminological (actually semantic) difference.

There is an identity for a reputation (state that persists before connections) -- and there is an "identity" as a "second factor" to ensure that there is no forgery of some network addresses controlled by authorities (DNS, NATs etc).

For sure I am not arguing about the first case (reputation, permanent identity).

The point I am trying to address is that bitcoin P2P protocol probably shouldn't allow connecting to remote peers by DNS or IP addresses without ensuring that the peer behind that address is the peer which was advertised via P2P. And static public key servers that purpose (and probably it is incorrect to call it "permanent identity", but its validation against an expected value is still an authentication).

**Comment by dr-orlovsky** [comment]
I'd like to understand better why you think that P2P network should have a quality that for automatic connections, every node is equally good (i.e. the rational behind that). I am not trying to persuade you, I am asking for better rationale - I think others may have similar questions and it will be nice to clarify the approach better (including the text of the document).

From my understanding, Bitcoin P2P network has a quality of eclipse attack. Connecting to peers which are guaranteed to be the peers known to be well-connected - or outside of the scope of a vulnerable network segment (for instance being in US for a Chinese node) is a way to mitigate that risks. If these nodes are not authenticated to be the nodes a peer tried to connect to (i.e. proving their network address was not stolen and they are just fakes run by a Chinese government) would make eclipse attacks more probable/simple. Or you mean that such nodes would be authenticated as a layer above this spec, while all other nodes ("automatically connected") can be a fakes without increasing chances of eclipse?

**Comment by yojoe** [comment]
> I'd like to understand better why you think that P2P network should have a quality that for automatic connections, every node is equally good (i.e. the rational behind that).

Bitcoin is special among P2P networks in regards to eclipse attacks as they can be easily/cheaply detected. All you need to do is periodically check the chain tip hash via other low-bandwidth, out-of-band channel(s). Either manually or you script it. It would be really hard for an adversary to eclipse all your communication channels and stop you from learning that another chain tip exists. Alternatively, you can already manually and explicitly add a few trusted "nodes" to the mix, e.g. Tor, I2P or CJDNS, whose address schemes are all "public key" based and don't rely on centralised authorities like DNS. So if you worry about eclipse attacks, you can already detect and somewhat counter them.

But the main point is that "public key based identities" and "authenticated connections" are not the silver bullet against eclipse attacks in **TRUSTLESS, PUBLIC, OPEN NETWORKS**.

> Connecting to peers which are guaranteed to be the peers known to be well-connected

That would be a "web of trust" and not a "trustless" topology.

> or outside of the scope of a vulnerable network segment (for instance being in US for a Chinese node) is a way to mitigate that risk

AFAIK that's exactly what the clearnet (IP4/6) peer selection algorithm already considers.

**Review by dr-orlovsky — APPROVED** [review]
I've got all answers to my questions, thank you. ACK

**Comment by kallewoof** [comment]
I don't see any un-addressed concerns regarding this pull request and will merge this soon unless someone pipes up. Thanks.

**Comment by dhruv** [comment]
@kallewoof we have a couple small changes coming up

**Comment by kallewoof** [comment]
It's helpful if you change to draft when you don't want it merged so maintainers know to skip over it.

**Comment by dhruv** [comment]
@kallewoof I'm sorry about that -- I was unclear about the workflow on the bips repo but I understand now. I have pushed up the latest changes and the working group (@sipa, @real-or-random and I) agrees it's rfm. 

Latest push:
- New rekeying mechanism that uses ChaCha20 keystream bytes and does not rely on SHA256 - reducing one dependency for the BIP324 requirements as well as a small efficiency gain.
- Changed rekey interval to 224 messages instead of 256 messages.
- Erlay message types in the message type id table
- Reference code changes for terminator derivation
- Changes to footnotes/rationale
- Grammar/spelling

**Comment by kallewoof** [comment]
No worries!

Maybe mark BIP-151 as "Replaced" in `README.mediawiki`, similar to BIP-1 replaced by BIP-2?

**Comment by dhruv** [comment]
> Maybe mark BIP-151 as "Replaced" in `README.mediawiki`, similar to BIP-1 replaced by BIP-2?

@kallewoof Done.

File: bip-0324.mediawiki
Comment (ajtowns) [reply bip-0324.mediawiki:523]: Finally got an implementation of this. Takes ~264 bytes of data per-peer to track the settings, and a `SET324ID` (that takes a `vector<pair<uint8_t, string>>` to define the mapping) with all the message types we currently support receiving is a 380 byte message. https://github.com/ajtowns/bitcoin/commits/202511-bip324-id/

**Pull request merged**
